# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [19]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [20]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [21]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [22]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [23]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [24]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [25]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [26]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [27]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [28]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [29]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [30]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [31]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [32]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [33]:
data.duplicated().sum()

71

In [34]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [35]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [36]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [37]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [38]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [39]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [24]:
data.groupby('children')['debt'].mean()

children
0    0.075438
1    0.092346
2    0.094542
3    0.081818
4    0.097561
5    0.000000
Name: debt, dtype: float64

In [26]:
data['children'].value_counts()

0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: children, dtype: int64

**Вывод:** Сравнивая между собой бездетные группы и группы с детьми, можно сделать вывод, что отсутсвие детей в семье уменьшает вероятность просрочки платежа. Однако можем увидеть группу с пятью детьми в семье. Задолженности у такой группы нет. Однако выборка слишком мала, чтобы делать какие-либо выводы. То же самое можно сказать и про группу с 4 детьми

In [40]:
children_pivot = data.pivot_table('debt', index='children', aggfunc=['sum', 'count', lambda x: '{:.2%}'.format(x[:].mean())],)
children_pivot = children_pivot.rename(columns={"sum":"количество должников", "count":"количество заемщиков", "<lambda>": "доля должников", "debt": "" })
children_pivot.index.name = 'количество детей'
children_pivot

количество должников количество заемщиков доля должников
                                                                         
количество детей                                                         
0                                1063                14091          7.54%
1                                 444                 4808          9.23%
2                                 194                 2052          9.45%
3                                  27                  330          8.18%
4                                   4                   41          9.76%
5                                   0                    9          0.00%

Вывод: Сравнивая между собой бездетные группы и группы с детьми, можно сделать вывод, что отсутсвие детей в семье уменьшает вероятность просрочки платежа. Однако можем увидеть группу с пятью детьми в семье. Задолженности у такой группы нет. Однако выборка слишком мала, чтобы делать какие-либо выводы. То же самое можно сказать и про группу с 4 детьми

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [27]:
data.groupby('family_status')['debt'].mean()

family_status
Не женат / не замужем    0.097639
в разводе                0.070648
вдовец / вдова           0.066246
гражданский брак         0.093130
женат / замужем          0.075606
Name: debt, dtype: float64

In [28]:
data['family_status'].value_counts()

женат / замужем          12261
гражданский брак          4134
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
Name: family_status, dtype: int64

**Вывод:** Пройдемся по группам:
1) Те, кто овдовел, имеют наименьший процент возврата долга. Вероятнее всего это связано с ответственностью за свою жизнь в отсутсвие любимого человека
2) Неплохой процент имеют люди, находящиеся в разводе или официальном браке. Возможно это связано с общей заботой о детях.
3) Самый худший процент у людей, находящихся в гражданском браке, а также у групп лиц, никогда не бывавших в браке. 

In [41]:
family_pivot = data.pivot_table('debt', index='family_status', aggfunc=['sum', 'count', lambda x: '{:.2%}'.format(x[:].mean())],)
family_pivot = family_pivot.rename(columns={"sum":"количество должников", "count":"количество заемщиков", "<lambda>": "доля должников", "debt": "" })
family_pivot.index.name = 'семейное положение'
family_pivot  

количество должников количество заемщиков доля должников
                                                                              
семейное положение                                                            
Не женат / не замужем                  273                 2796          9.76%
в разводе                               84                 1189          7.06%
вдовец / вдова                          63                  951          6.62%
гражданский брак                       385                 4134          9.31%
женат / замужем                        927                12261          7.56%

**Вывод:** 1) Те, кто овдовел, имеют наименьший процент возврата долга. Вероятнее всего это связано с ответственностью за свою жизнь в отсутсвие любимого человека 2) Неплохой процент имеют люди, находящиеся в разводе или официальном браке. Возможно это связано с общей заботой о детях. 3) Самый худший процент у людей, находящихся в гражданском браке, а также у групп лиц, никогда не бывавших в браке.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Для упрощения возьмем столбец total_income_category, созданный в задании 16, где
0–30000 — 'E';
30001–50000 — 'D';
50001–200000 — 'C';
200001–1000000 — 'B';
1000001 и выше — 'A'.

In [37]:
data.groupby('total_income_category')['debt'].mean()

total_income_category
A    0.080000
B    0.070602
C    0.084982
D    0.060172
E    0.090909
Name: debt, dtype: float64

In [38]:
data['total_income_category'].value_counts()

C    15921
B     5014
D      349
A       25
E       22
Name: total_income_category, dtype: int64

**Вывод:** Группы A и E слишком малы для какого-либо вывода. То же самое можно сказать и о группе D. Уровень дохода у группы C ниже, чем у группы B, и процент возврата кредита выше у вторых. Но с учетом сравнения лишь двух групп не могу сделать вывод о зависимости между уровнем дохода и возвратом кредита в срок.


In [48]:
def mean_isl(x):
    return '{:.2%} '.format(x.mean())
data.groupby(['total_income_category'])['debt'].agg(['count', mean_isl]).sort_values('mean_isl')

count mean_isl
total_income_category                
D                        349   6.02% 
B                       5014   7.06% 
A                         25   8.00% 
C                      15921   8.50% 
E                         22   9.09%

**Вывод:** Группы A и E слишком малы для какого-либо вывода. То же самое можно сказать и о группе D. Хоть и уровень дохода у группы C ниже, чем у группы B, и процент возврата кредита выше у вторых. Но с учетом сравнения лишь двух групп не могу сделать вывод о зависимости между уровнем дохода и возвратом кредита в срок.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [35]:
data.groupby('purpose_category')['debt'].mean()

purpose_category
операции с автомобилем      0.093480
операции с недвижимостью    0.072551
получение образования       0.092528
проведение свадьбы          0.079118
Name: debt, dtype: float64

In [36]:
data['purpose_category'].value_counts()

операции с недвижимостью    10751
операции с автомобилем       4279
получение образования        3988
проведение свадьбы           2313
Name: purpose_category, dtype: int64

**Вывод:** Хуже всего обстоят дела у тех, кто берет кредит на образование и покупку автомобиля. Первые, возможно, являются студентами и им сложнее отдавать долг. Получше отдают кредит те, кто покупает жилье или празднует свадьбу. По опыту своих друзей могу сказать, что кредиты отбиваются подарками от гостей свадьбы. 

In [49]:
data.groupby(['purpose_category'])['debt'].agg(['count', mean_isl]).sort_values('mean_isl')

count mean_isl
purpose_category                        
операции с недвижимостью  10751   7.26% 
проведение свадьбы         2313   7.91% 
получение образования      3988   9.25% 
операции с автомобилем     4279   9.35%

**Вывод:** Хуже всего обстоят дела у тех, кто берет кредит на образование и покупку автомобиля. Первые, возможно, являются студентами и им сложнее отдавать долг. Получше отдают кредит те, кто покупает жилье или празднует свадьбу. По опыту своих друзей могу сказать, что кредиты отбиваются подарками от гостей свадьбы.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

**Ответ:** Либо случайно, либо по причине того, что это безработные (можно посмотреть на вывод первых 20 строк в датафрейме и увидеть, что есть пропуски в уровне дохода)

Возможно кредит можно взять и без указания этой информации

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

**Ответ:** Медианные значения не так чувствительны для количественных переменных. Если есть значения, которые сильно выбиваются из среднего значения

### Шаг 4: общий вывод.

В ходе проведения исследования можно сделать несколько выводов:
1. Чем меньше детей - тем выше процент возврата
2. Семейное положения также влияет на возврат кредита в срок
3. Если судить по уровню дохода, то нельзя сделать однозначных выводов. По той выборке, что была представлена, уровень дохода действительно влияет (чем выше доход - тем ниже процент просрочки). Но нужно больше данных по другим уровням дохода
4. Цели для кредита тоже влияют. Особенно страдают студенты и люди, покупающие автомобиль

1. Так самыми надежными заемщиками являются те кто вообще не имеет детей, они не возвращают кредит всего в 7.54%, в отличие от клиентов имеющих двое детей, которые не так добросовестно относятся к задолженности и не возвращают в 9.45% случаев.
2. Есть ли зависимость между семейным положением и возвратом кредита в срок? Да, есть, семейное положение влияет на возврат кредита в срок. Заемщики, которые имеют статус "не женат/не замужем" имеют самый высокий процент невозврата. В отличие от людей, повтерявших своих спутников жизни, записанные в колонку "вдовец / вдова". Такие люди как правило более ответственные за свою жизнь. Плюс у них могут быть уже взрослые дети, которые могут сами себя обеспечить
3. По целям самые добросовестные заемщики те кто берет деньги на недвижимость, они не выплачивают в 7.26% случаев. Заёмщики берущие деньги на автомобиль не платят в 9.35% случаев.
4. Есть ли зависимость между уровнем дохода и возвратом кредита в срок? Точно сказать нельзя, ибо не хватает данных для друггих категорий.

Однако можно сделать некий портрет хорошего заемщика - **вдова/вдовец без детей с уровнем дохода свыше 200001 и ниже 1000000, который/которая берет кредит на недвижимость.**

Худший - **не женат/не замужем с двумя детьми с доходом 50001-200000, берущие кредит на автомобиль**
